bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/messages/ParadoxIPPacket.java

```diff
@@ -0,0 +1,191 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.paradoxalarm.internal.communication.messages;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import org.openhab.binding.paradoxalarm.internal.util.ParadoxUtil;
+
+/**
+ * The {@link ParadoxIPPacket} This class is object representing a full IP request packet. Header and payload together.
+ *
+ * @author Konstantin_Polihronov - Initial contribution
+ */
+public class ParadoxIPPacket {
+
+    public static final byte[] EMPTY_PAYLOAD = new byte[0];
+
+    /**
+     * Start of header - always 0xAA
+     */
+    private byte startOfHeader = (byte) 0xAA;
+
+    /**
+     * Payload length - 2 bytes (LL HH)
+     */
+    private short payloadLength = 0;
+
+    /**
+     * "Message Type: 0x01: IP responses 0x02: Serial/pass through cmd response
+     * 0x03: IP requests 0x04: Serial/pass through cmd requests"
+     */
+    private byte messageType = 0x03;
+
+    /**
+     * "IP Encryption 0x08: Disabled 0x09: Enabled"
+     */
+    private byte encryption = 0x08;
+    private byte command = 0;
+    private byte subCommand = 0;
+    private byte unknown0 = 0x0A;
+
+    /**
+     * Padding bytes to fill the header to 16 bytes with 0xEE.
+     */
+    private long theRest = 0xEEEEEEEEEEEEEEEEl;
+    private byte[] payload;
+    private boolean isChecksumRequired;
+
+    public ParadoxIPPacket(IPPacketPayload payload) throws IOException {
+        this(payload.getBytes(), true);
+    }
+
+    public ParadoxIPPacket(String payload, boolean isChecksumRequired) throws IOException {
+        this(payload.getBytes("US-ASCII"), isChecksumRequired);
+    }
+
+    public ParadoxIPPacket(byte[] payload, boolean isChecksumRequired) throws IOException {
+        this.isChecksumRequired = isChecksumRequired;
+
+        if (payload == null) {
+            this.payload = new byte[0];
+            this.payloadLength = 0;
+        } else {
+            this.payload = payload;
+            this.payloadLength = (short) payload.length;
+        }
+
+        // TODO: Figure out how to fill up to 16, 32, 48, etc sizes with 0xEE

```

Comment:
There is a TODO, what should be done and can it be completed before merging the PR?

Reply:
This is not mandatory feature but it's good to implement it as per reverse engineering document by the guy who initially did the RE.